### PR TITLE
Add buffer placement pass to inject dellocs

### PR DESCRIPTION
**Pull request by flaub**

### CODE REVIEW

File: pmlc/dialect/pxa/transforms/buffer_placement.cc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+struct BufferPlacementPass : public BufferPlacementBase<BufferPlacementPass> {
+  void runOnFunction() final {
+    auto fn = getFunction();
+    fn.walk([&](AllocOp alloc) {
+      IVLOG(1, "alloc: " << debugString(*alloc));
+
+      Operation *lastOp = alloc;
+      auto uses = getIndirectUses(alloc);
+      auto itUse = uses.begin();
+      while (itUse != uses.end()) {
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/buffer_placement.cc:36]: Can we just use a `for (auto& use : uses)` here?

File: pmlc/dialect/pxa/transforms/buffer_placement.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        if (!use->isBeforeInBlock(lastOp)) {
+          lastOp = use;
+        }
+
+        ++itUse;
+      }
+      IVLOG(1, "  last: " << debugString(*lastOp));
+
+      if (isa<ReturnOp>(lastOp)) {
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/buffer_placement.cc:55]: You could check this in the loop above and exit early.

File: pmlc/dialect/pxa/transforms/buffer_placement.cc (hunk 1 above)
Comment (adstraw) [reply pmlc/dialect/pxa/transforms/buffer_placement.cc]: Done.

File: pmlc/dialect/pxa/transforms/buffer_placement.cc (hunk 2 above)
Comment (adstraw) [reply pmlc/dialect/pxa/transforms/buffer_placement.cc]: Done.

File: pmlc/dialect/pxa/transforms/buffer_placement.cc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    fn.walk([&](AllocOp alloc) {
+      IVLOG(3, "alloc: " << debugString(*alloc));
+
+      Operation *lastOp = alloc;
+      Block *allocBlock = alloc.getOperation()->getBlock();
+      for (auto &itUse : getIndirectUses(alloc)) {
+        Operation *use = itUse.getOwner();
+        IVLOG(3, "  use: " << debugString(*use));
+
+        while (use->getBlock() != allocBlock) {
```
Comment (jbruestle) [inline pmlc/dialect/pxa/transforms/buffer_placement.cc:27]: You can replace this loop with:
use = allocBlock->findAncestorOpInBlock(use);
assert(use && "no common ancestor)

Comment (adstraw) [reply pmlc/dialect/pxa/transforms/buffer_placement.cc]: Done.

**Pull request merged**